A constrained TypeVar value fails type() narrowing tests

**Bug Report**

When a `TypeVar` is constrained, passing a value of the constrained type into `type()` fails narrowing tests.

**To Reproduce**

```python
from typing import TypeVar

W = TypeVar("W", int, str)

def fn(w: W) -> W:
    if type(w) is str:
        # main.py:30: note: Revealed type is "builtins.int"
        # main.py:30: note: Revealed type is "builtins.str"
        reveal_type(w)

        # reports error:
        # main.py:31: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]
        s: str = w
        print("s: %s" % s)

    elif type(w) is int:
        # same in this block
        reveal_type(w)
        i: int = w
        print("s: %d" % i)
        
    return w
```
If I change `W` to use binding, the problem goes away. That is, if I use this, no errors are reported.

```python
W = TypeVar("W", bound=int|str)
```

Testing the original code in Pyright does not report any problems.

Using `isinstance()` makes it work for class types in general, but cannot be used against `TypedDict` instances and where a specific class type is required.

**Expected Behavior**

The condition `type(w) is str` should narrow `w` to have `str` in its conditional block. 

**Actual Behavior**

These errors are reported:

> main.py:35: error: Incompatible types in assignment (expression has type "str", variable has type "int")  [assignment]
> main.py:44: error: Incompatible types in assignment (expression has type "int", variable has type "str")  [assignment]

**Your Environment**

This is not environment specific and I can reproduce the same on Windows or Linux.

- Mypy version used: 1.14.1
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -5914,6 +5914,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
         def combine_maps(list_maps: list[TypeMap]) -> TypeMap:
             """Combine all typemaps in list_maps into one typemap"""
+            if all(m is None for m in list_maps):
+                return None
             result_map = {}
             for d in list_maps:
                 if d is not None:

--- a/test-data/unit/check-isinstance.test
+++ b/test-data/unit/check-isinstance.test
@@ -2661,6 +2661,33 @@ y: Any
 if type(y) is int:
     reveal_type(y) # N: Revealed type is "builtins.int"
 
+[case testTypeEqualsCheckUsingIsNonOverlapping]
+# flags: --warn-unreachable
+from typing import Union
+
+y: str
+if type(y) is int:  # E: Subclass of "str" and "int" cannot exist: would have incompatible method signatures
+    y # E: Statement is unreachable
+else:
+    reveal_type(y) # N: Revealed type is "builtins.str"
+[builtins fixtures/isinstance.pyi]
+
+[case testTypeEqualsCheckUsingIsNonOverlappingChild-xfail]
+# flags: --warn-unreachable
+from typing import Union
+
+class A: ...
+class B: ...
+class C(A): ...
+x: Union[B, C]
+# C instance cannot be exactly its parent A, we need reversed subtyping relationship
+# here (type(parent) is Child).
+if type(x) is A:
+    reveal_type(x)  # E: Statement is unreachable
+else:
+    reveal_type(x)  # N: Revealed type is "Union[__main__.B, __main__.C]"
+[builtins fixtures/isinstance.pyi]
+
 [case testTypeEqualsNarrowingUnionWithElse]
 from typing import Union
 

--- a/test-data/unit/check-typevar-values.test
+++ b/test-data/unit/check-typevar-values.test
@@ -744,5 +744,4 @@ def fn(w: W) -> W:
     elif type(w) is int:
         reveal_type(w)  # N: Revealed type is "builtins.int"
     return w
-
 [builtins fixtures/isinstance.pyi]

--- a/test-data/unit/check-typevar-values.test
+++ b/test-data/unit/check-typevar-values.test
@@ -732,3 +732,17 @@ def foo3(x: NT) -> None:
 def foo4(x: NT) -> None:
     p, q = 1, 2.0  # type: (int, float)
 [builtins fixtures/tuple.pyi]
+
+[case testTypeVarValuesNarrowing]
+from typing import TypeVar
+
+W = TypeVar("W", int, str)
+
+def fn(w: W) -> W:
+    if type(w) is str:
+        reveal_type(w)  # N: Revealed type is "builtins.str"
+    elif type(w) is int:
+        reveal_type(w)  # N: Revealed type is "builtins.int"
+    return w
+
+[builtins fixtures/isinstance.pyi]

